### ISSUE TITLE
FR: In recovery package, pass context to custom recovery handler

### ISSUE THREAD

**Issue by jgiles** [grpc-ecosystem/go-grpc-middleware#168]
Currently, a custom recovery handler looks like:

```
type RecoveryHandlerFunc func(p interface{}) (err error)
```

see https://github.com/grpc-ecosystem/go-grpc-middleware/blob/498ae206fc3cfe81cd82e48c1d4354026fa5f9ec/recovery/interceptors.go#L13

It would be useful if the recovery function could access the request context - in particular, to access tags attached by `grpc_ctxtags` or other context-attached values. The context is available in both unary and stream server interceptors (via `stream.Context`).

For backwards compatibility, it is probably infeasible to change the signature of the custom handler. However, we could introduce

```
type CtxRecoveryHandlerFunc func(ctx context.Context, p interface{}) (err error)
```

and have an alternative option for setting it.

**Comment by hxzhao527** [comment]
arguments in `grpc.UnaryServerInterceptor` except `grpc.UnaryHandler` maybe also useful.
Same with `grpc.StreamServerInterceptor`

**Comment by johanbrandhorst** [comment]
Definitely want this, the context may contain important tracing and metrics metadata if you're using OpenCensus which stores tags in the context.

**Comment by hxzhao527** [comment]
Actullay, user can pass `context.Context` through `panic` by themself.
In service impl:
```go
type RPCError struct {
	Message string
	Payload interface{}
	Ctx context.Context
}
```
```go
// TryPanic will just panic.
func (s *Server) TryPanic(ctx context.Context, in *empty.Empty) (*empty.Empty, error) {
	panic(RPCError{Message:"just trying panic and recovery", Ctx:ctx, Payload: in})
}
```
In `RecoveryHandlerFunc`
```go
func GlobalRecoveryHandler(p interface{}) error {
	switch err := p.(type) {
	case string:
		return status.Errorf(codes.Internal, "%s", err)
	case RPCError:
		return status.Errorf(codes.Internal, "%#v", p)
	}
	return status.Errorf(codes.Internal, "%#v", p)
}
```
use type-switch and structured error works well.
The complete demo is [here](https://github.com/hxzhao527/grpcdemo/blob/master/pkg/server/recover.go#L51).

**Comment by johanbrandhorst** [comment]
Implying that you should expect to have control over the contents of your panic is completely ridiculous.

**Comment by hxzhao527** [comment]
Not really. Controling over the content of `panic`  just is a way to format your `error`, just like the example in [Golang wiki](https://github.com/golang/go/wiki/PanicAndRecover#usage-in-a-package).
The above code is just my Implementation method in a little demo. I just want to show how pass `ctx` in the existing way.

In my opinion, passing `ctx` is infructuous. `ctx` doesn't take the entire information of the rpc call. To get the full stack of rpc call, user need to deal with `ctx` by themself. But now there is no any API to do that.

If we must add a new api to get the info of rpc call, passing a new struct that contains all info maybe better. Otherwise we need to add a new api later.
